client/webserver/http.go

```diff
@@ -190,6 +197,109 @@ func (s *WebServer) handleOrders(w http.ResponseWriter, r *http.Request) {
+	filter.Hosts = r.Form["hosts"]
+	assets := r.Form["assets"]
+	filter.Assets = make([]uint32, len(assets))
+	for k, assetStrId := range assets {
+		assetNumId, err := strconv.Atoi(assetStrId)
+		if err != nil {
+			log.Errorf("error parsing asset id: %v", err)
+			http.Error(w, http.StatusText(http.StatusBadRequest), http.StatusBadRequest)
+			return
+		}
+		filter.Assets[k] = uint32(assetNumId)
+	}
+	statuses := r.Form["statuses"]
+	filter.Statuses = make([]order.OrderStatus, len(statuses))
+	for k, statusStrId := range statuses {
+		statusNumId, err := strconv.Atoi(statusStrId)
+		if err != nil {
+			http.Error(w, http.StatusText(http.StatusBadRequest), http.StatusBadRequest)
+			log.Errorf("error parsing status id: %v", err)
+			return
+		}
+		filter.Statuses[k] = order.OrderStatus(statusNumId)
+	}
+
+	ords, err := s.core.Orders(filter)
+	if err != nil {
+		log.Errorf("error retrieving order: %v", err)
+		http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
+		return
+	}
+
+	var buffer bytes.Buffer
+	csvWriter := csv.NewWriter(&buffer)
+	// UseCRLF for Window users
+	csvWriter.UseCRLF = strings.Contains(r.UserAgent(), "Windows")
+
+	err = csvWriter.Write([]string{
+		"Host",
+		"Base",
+		"Quote",
+		"Ask",
+		"Offer",
+		"Type",
+		"Status",
+		"Rate",
+		"Filled (%)",
+		"Settled (%)",
+		"Quantity",
+		"Time",
+	})
+	if err != nil {
+		log.Errorf("error writing CSV: %v", err)
+		http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
+		return
+	}
+	for _, ord := range ords {
+		ordReader := orderReader{ord}
+
+		srvTime := encode.UnixTimeMilli(int64(ord.Stamp))
+		err = csvWriter.Write([]string{
+			ord.Host,                                  // Host
+			ord.BaseSymbol,                            // Base
+			ord.QuoteSymbol,                           // Quote
+			ordReader.AskString(),                     // Ask
+			ordReader.OfferString(),                   // Offer
+			ordReader.TypeString(),                    // Type
+			ordReader.StatusString(),                  // Status
+			ordReader.RateString(),                    // Rate
+			ordReader.FilledPercent(),                 // Filled
+			ordReader.SettledPercent(),                // Settled
+			ordReader.OfferString(),                   // Quantity
+			srvTime.Format("02/01/2006, 03:04:05 PM"), // Time
+		})
+		if err != nil {
+			log.Errorf("error writing CSV: %v", err)
+			http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
+			return
+		}
+	}
+	csvWriter.Flush()
+	data := buffer.Bytes()
+	w.Header().Set("Content-Disposition", "attachment; filename=orders.csv")
+	w.Header().Set("Content-Type", "text/csv")
+	w.Header().Set("Content-Length", fmt.Sprintf("%d", len(data)))
+
+	_, err = w.Write(data)
+	if err != nil {
+		log.Errorf("error writing to client body: %v", err)
+		http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
```

Comment:
At this point if there is an error from `http.ResponseWriter.Write`, it's too late to call `http.Error` (already written the header).  Just log an let the function return.

Reply:
Thank you